system-addon/lib/NewTabInit.jsm

```diff
@@ -37,6 +37,11 @@ this.NewTabInit = class NewTabInit {
           this._queue.clear();
         }
         break;
+      case at.SAVE_SESSION_PERF_DATA:
+        if (action.data.topsites_first_painted_ts && action._target.url === "about:home") {
```

Comment:
I wanted to use `visibility_event_rcvd_ts` but that one isn't coming through for me for the first `about:home` on browser startup.

Reply:
So... the things to consider is that first_painted might happen before visibility if preloaded, but we only have preloaded `about:newtab`s and not `about:home` (at least for now). Also this assumes search gets rendered before topsites, which might be okay to assume for now?

I'm guessing you checked what possible actions were already being sent on early loads (running into #3455).

Otherwise we would need to detect if visible on mount to dispatch an action. And if not visible yet wait until then..

Reply:
Ideally, we would have content announce when things are happening and different feeds can take appropriate action. But right now it's written a bit more imperatively.